checkbox-ng/plainbox/impl/session/README.md

```diff
@@ -0,0 +1,78 @@
+# System Information Collection
+
+Checkbox collects every information on the system it can using Collectors.
+A collector is a tool that can provide a JSON-serializable set of information.
+When Checkbox starts, it will leverage these collectors and collect/store
+all information it can gather in the session storage before the first
+session checkpoint is created. From then onward these information will remain
+in memory (and on disk in the checkpoint). When generating a submission report
+checkbox will include all information in a top level field of the json called
+"system_information".
+
+## Format
+
+A collector can either run succesfully or fail. Regardless of the result,
+running a collector will create a new field in the submission file following
+this format:
+
+```
+"system_information" : {
+  "collector_name" : {
+    "tool_version" : collector_version,
+    "success" : True/False,
+    "outputs" : { ... }
+}
+```
+
+The outputs field's format depends on the success of the collection.
+
+If it ran successfully, the output field will have the follwing structure:
+
+```
+  "outputs" : {
+    "json_output" : collector_json_output,
```

Comment:
I still struggle to understand why we need this dynamic schema.

The whole "outputs" shenanigans looks unnecessary. Two streams are produced by a collector, stdout, and stderr. If the collector run successfully, the stdout has the collection json. It's not like the collector will open different streams if it's unsuccessful.

Reply:
I think we are going back to the initial design of always reporting `stdout` and `stderr`. I don't want that. I am convinced that we want the output json to be interpreted so that we know it is valid and we can actually use it in checkbox as well in the future. We went back and fourth on this a few times already. Can you write an example output that you would like to see? Something like
```json
"system_information" : {
    "tool_success" : {
        "version" : "1",
        "success" : true,
        "outputs" : { $SUGGESTION_SUCCESS }
    },
    "tool_failure" : {
        "version" : "1",
        "success" : false,
        "outputs" : { $SUGGESTION_FAILURE }
    }
}
```

Reply:
Isn't the information replicated. I mean "success: true" is redundant with "tool_success".

Reply:
(tool success is the name of the tool, fake it like tool success is "inxi" and tool failure is "false")

Reply:
Gotcha. So like a "successful_tool".

My example:
```json
"system_information" : {
    "happy_inxi" : {
        "version" : "1",
        "retcode" : 0,
        "stdout": "{\"Computer\": \"yes\", \"Brand\": \"ACME\", \"RAM\": \"A lot\"}"
        "stderr": ""
    },
    "unhappy_lshw" : {
        "version" : "1",
        "retcode" : 1,
        "stdout": ""
        "stderr": "Can't open /sys/some/info. Operation not permited"
    },
    "unhappy_another": {
        "version": "1",
        "retcode": 2,
        "stdout" : "{\"first field\": \"crashed while printing, bad json",
        "stderr": "Ran out of memory, omg"
    }
}
```

Reply:
This way the json is unparsed and it is one of the things we want to avoid in the happy case because it makes the schema less descriptive (in case success is true, you can parse stdout and it will be a valid json). This was one of the first requirement that was put in place, when we are in the happy case, there is no string parsing to be done, the data is already parsed in a usable form. 

That is the reason why I originally proposed to have a stdout field that could either be a string or a dict, that was struck down because we don't want a union there. The final iteration of this loop was to have a `json_output` in the happy case that contained the parsed result and a failure output (in whatever form, just for debug/presentation purposes) in the unhappy case

Reply:
As per our dicussione we have decided to (via meet):
- Change `json_output` -> `payload`, as it is a better name that both signals the meaning and the type better
- Change the junja template to contain a version
- Change the jinja template to not encode the json manually
- Change the "system_information" field to have a version

Reply:
Implemented all changes